### PR TITLE
!!! TASK: Separate `ForwardException` from `StopActionException`

### PR DESCRIPTION
**Pull request by mhsdesign**
My two cents to https://github.com/neos/flow-development-collection/pull/3232

tests ran here https://github.com/neos/flow-development-collection/pull/3281

### CODE REVIEW

File: Neos.Flow/Classes/Mvc/Exception/StopActionException.php [hunk 1]
```diff
@@ -23,7 +23,24 @@
...
+    private function __construct(string $message, int $code, ?\Throwable $previous, ActionResponse $response)
+    {
+        parent::__construct($message, $code, $previous);
+        $this->response = $response;
+    }
+
+    public static function create(
+        ActionResponse $response,
+        string $message,
+        int $code
```
Comment (mhsdesign) [inline Neos.Flow/Classes/Mvc/Exception/StopActionException.php]: im unsure _why_ there even has to be a message and a code?

Isnt this purely for control flow?

Comment (kitsunet) [reply Neos.Flow/Classes/Mvc/Exception/StopActionException.php]: nod, but then that's what exceptions are for? :D

Comment (mhsdesign) [reply Neos.Flow/Classes/Mvc/Exception/StopActionException.php]: it _could_ extract the message from the request on its own.

```php
$infoMessage = sprintf(
    '%s %s',
    $response->getStatusCode(),
    ResponseInformationHelper::getStatusMessageByCode($response->getStatusCode())
);
```

But then again we are used to specifying a message (wich will be used for the response content as well)

https://github.com/neos/flow-development-collection/blob/a5cd6e34e777a0e3658d5ef4bb9c1528381fcb04/Neos.Flow/Classes/Mvc/Controller/RestController.php#L68

So lets keep the `$infoExceptionMessage` only for informational purposes, but make it optional.
I would fix the exception timestamp code to a fix number though.

Comment (kitsunet) [reply Neos.Flow/Classes/Mvc/Exception/StopActionException.php]: Fine

File: Neos.Flow/Classes/Mvc/Exception/ForwardException.php [hunk 2]
```diff
@@ -19,32 +19,21 @@
...
         $this->nextRequest = $nextRequest;
     }
 
-    /**
-     * Returns the next request
-     *
-     * @return ActionRequest
-     */
-    public function getNextRequest()
+    public static function create(ActionRequest $nextRequest)
```
Comment (kitsunet) [inline Neos.Flow/Classes/Mvc/Exception/ForwardException.php]: then maybe same signature here? with request and exceptionmessage?

Comment (mhsdesign) [reply Neos.Flow/Classes/Mvc/Exception/ForwardException.php]: Yes added a $details param for the ForwardException as well.

Comment (mhsdesign) [reply Neos.Flow/Classes/Mvc/Exception/ForwardException.php]: Though we dont use this param yet in `forwardToRequest`

File: Neos.Flow/Classes/Mvc/Exception/StopActionException.php [hunk 3]
```diff
@@ -19,11 +19,38 @@
...
+    {
+        parent::__construct($message, $code, $previous);
+        $this->response = $response;
+    }
+
+    /**
+     * @param ActionResponse $response The response to be received by the MVC Dispatcher.
+     * @param string $details Additional details just for this exception, in case it is logged (the regular exception message).
+     */
+    public static function createForResponse(ActionResponse $response, string $details): self
```
Comment (mhsdesign) [inline Neos.Flow/Classes/Mvc/Exception/StopActionException.php:46]: or `createToReturnResponse` or `createWithResponseToBeReturned`?

And `createToHandleNextRequest` respectively?

Comment (mhsdesign) [reply Neos.Flow/Classes/Mvc/Exception/StopActionException.php:46]: Maybe we can make the factories internal? Then it wouldnt matter that much ^^

Comment (kitsunet) [reply Neos.Flow/Classes/Mvc/Exception/StopActionException.php:46]: it IS the api to trigger this behavior though, so technically this should be API.

Comment (mhsdesign) [reply Neos.Flow/Classes/Mvc/Exception/StopActionException.php:46]: yes i agree, so do you like the namings or any of my suggestion better?

**Review by kitsunet — APPROVED** [review]

**Pull request merged**